MapboxCoreNavigation/FreeDriveLocationManager.swift

```diff
@@ -0,0 +1,156 @@
+        self.directions = directions
+        
+        let settingsProfile = SettingsProfile(application: ProfileApplication.kMobile, platform: ProfilePlatform.KIOS)
+        let navigator = Navigator(profile: settingsProfile, config: NavigatorConfig() , customConfig: "")
+        
+        let proxyDelegate = ProxyDelegate()
+        proxyDelegate.navNative = navigator
+        self.proxyDelegate = proxyDelegate
+        
+        super.init()
+
+        super.delegate = proxyDelegate
+    }
+    
+    public override func startUpdatingLocation() {
+        startUpdatingLocation(completionHandler: nil)
+    }
+    
+    /**
+     Starts the generation of location updates with an optional completion handler that gets called when the location manager is ready to receive snapped location updates.
+     */
+    public func startUpdatingLocation(completionHandler: ((Error?) -> Void)?) {
+        super.startUpdatingLocation()
+        
+        let tilesVersion = RouteTilesVersion(with: directions.credentials)
+        tilesVersion.getAvailableVersions { availableVersions in
+            if let latestVersion = availableVersions.last {
+                tilesVersion.currentVersion = latestVersion
+                do {
+                    try self.configureNavigator(withTilesVersion: latestVersion)
+                    completionHandler?(nil)
+                } catch {
+                    completionHandler?(error)
+                }
+            }
+        }
+    }
+    
+    func configureNavigator(withTilesVersion tilesVersion: String) throws {
+        let endpointConfig = TileEndpointConfiguration(directions: directions, tilesVersion: tilesVersion)
+
+        guard var tilesURL = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first else {
+            preconditionFailure("No Caches directory to create the tile directory inside")
+        }
+        tilesURL.appendPathComponent(tilesVersion, isDirectory: true)
+        // Tiles with different versions shouldn't be mixed, it may cause inappropriate Navigator's behaviour
+        try FileManager.default.createDirectory(at: tilesURL, withIntermediateDirectories: true, attributes: nil)
+        let params = RouterParams(tilesPath: tilesURL.path, inMemoryTileCache: nil, mapMatchingSpatialCache: nil, threadsCount: nil, endpointConfig: endpointConfig)
+        
+        proxyDelegate?.navNative?.configureRouter(for: params)
+    }
+
+    public func updateLocation(_ location: CLLocation?) {
+        guard let location = location else { return }
+        stopUpdatingLocation()
+        stopUpdatingHeading()
+        proxyDelegate?.locationManager(self, didUpdateLocations: [location])
+    }
+
+    override public var delegate: CLLocationManagerDelegate? {
+        get {
+            proxyDelegate?.delegate
+        }
+        set {
+            proxyDelegate?.delegate = newValue
+        }
+    }
+
+    private class ProxyDelegate: NSObject, CLLocationManagerDelegate {
```

Comment:
This is an unusual pattern in Swift, sort of reminiscent of PImpl but also redundant in its conformance to CLLocationManagerDelegate. I think the reason for this approach was the need to avoid making FreeDriveLocationManager both inherit from CLLocationManager and conform to CLLocationManagerDelegate. The inverse approach would be cleaner: FreeDriveLocationManager should continue to conform to CLLocationManagerDelegate directly but _contain_ an object that inherits from CLLocationManager.